### ISSUE TITLE
SIGBUS on OS X

### ISSUE THREAD

**Issue by reaperhulk** [pyca/cryptography#2503]
Replicating some info from https://github.com/pyca/pyopenssl/issues/391 into the correct issue tracker:

Cryptography binds functions that may or may not be available. Unfortunately, cffi provides us no method by which we can say "include this symbol if present, otherwise leave it out". Instead we need to call `cdef` to declare our types and functions, but we don't actually know if those functions exist until we invoke the compiler/linker. To handle this we developed a pattern whereby we use the C preprocessor to introspect our OpenSSL (via the headers) and, if symbols are missing, we declare them as null function pointers (or other values if they're not functions) and set a flag. During import of the Python module we then iterate over the set of conditional symbols and remove all attributes that aren't present from the `lib` object.

In this particular case we declared a flag `Cryptography_HAS_EC_NISTP_64_GCC_128` and then used it in [this conditional](https://github.com/pyca/cryptography/blob/master/src/_cffi_src/openssl/ec.py#L362-L371):

```cpp
#if defined(OPENSSL_NO_EC) || OPENSSL_VERSION_NUMBER < 0x1000100f || \
    defined(OPENSSL_NO_EC_NISTP_64_GCC_128)
static const long Cryptography_HAS_EC_NISTP_64_GCC_128 = 0;

const EC_METHOD *(*EC_GFp_nistp224_method)(void) = NULL;
const EC_METHOD *(*EC_GFp_nistp256_method)(void) = NULL;
const EC_METHOD *(*EC_GFp_nistp521_method)(void) = NULL;
#else
static const long Cryptography_HAS_EC_NISTP_64_GCC_128 = 1;
#endif
```

So far so good. The problem arises when building a universal library on OS X and then linking against the static version of it. To understand that we need a quick digression into what a universal library is on OS X...

Universal libraries are simply multiple CPU architectures crammed together in one file (typically using a tool called `lipo`). The linker simply looks at what arch it is linking for and uses those while ignoring the others. Universal libraries typically have i386 and x86_64 symbols in them these days, but just a few years ago you could find ppc, ppc64, i386, and x86_64 inside some libs.

When compiling cryptography (or any project) against a universal lib you specify the headers and then the library location (if they're not already in your include and lib search paths) and then the compiler does the rest. Unfortunately, universal libraries apparently assume that the headers will be the same between architectures. This is not true for any project that leaks build flags into the public headers. In OpenSSL's case this comes through with `OPENSSL_NO_EC_NISTP_64_GCC_128`. The above code declares `EC_GFp_nistp224_method` (and friends) as NULL function pointers if `OPENSSL_NO_EC_NISTP_64_GCC_128` is set. Unfortunately, while this is set on the i386 build, it is **not** set on x86_64. Homebrew, when building a universal library, chooses to keep the headers from the i386 build. So, when cryptography compiles the C preprocessor dutifully says that `OPENSSL_NO_EC_NISTP_64_GCC_128` is set and therefore `EC_GFp_nistp224_method`, etc are not present and should be defined as NULL function pointers.

Of course, this should still be caught right? After all, the `ec.h` header contains a different declaration for this function and we explicitly include the headers so that any mismatch will cause a compiler error! Let's look at the header file!

```cpp
# ifndef OPENSSL_NO_EC_NISTP_64_GCC_128
/** Returns 64-bit optimized methods for nistp224
 *  \return  EC_METHOD object
 */
const EC_METHOD *EC_GFp_nistp224_method(void);

/** Returns 64-bit optimized methods for nistp256
 *  \return  EC_METHOD object
 */
const EC_METHOD *EC_GFp_nistp256_method(void);

/** Returns 64-bit optimized methods for nistp521
 *  \return  EC_METHOD object
 */
const EC_METHOD *EC_GFp_nistp521_method(void);
# endif
```

:sob:

But wait, even if this all gets compiled the symbol is still present in the `libcrypto.a` right? Why isn't that called? I'm still not clear on that one but for now it's :sparkles:linker reasons:sparkles:. Similarly, this doesn't occur if you dynamically link rather than statically link: :sparkles:linker reasons:sparkles:.

Finally, if we're removing the function conditionally why is it being called and failing? Well, internally `EC_GROUP_new_by_curve_name` (and a few other functions) call it. Since the x86_64 build **was** compiled with `EC_NISTP_64_GCC_128` support they believe the function is available, but when they call it they get the null function pointer and it crashes. This is from the C layer, so the removal of the Python attribute is not relevant.

## Summary and What We're Gonna Do™

So, when running in 64-bit mode linked against a universal OpenSSL built using homebrew the current cryptography release (and master) will `SIGBUS` due to a header mismatch during compilation causing a symbol to be improperly defined as a NULL function pointer. This affects many scenarios involving elliptic curve keys on OS X right now, although the failure mode (crash!) is at least mostly safe.

The functions in question have no need to be bound at all (they're really internal implementation details and are of no benefit to us), so dropping them from the bindings will resolve this issue. Going forward, I'm going to propose altering our NULL function pointer approach to instead be something more like

```cpp
 const EC_METHOD *EC_GFp_nistp224_method(void) {
     printf("EC_GFp_nistp224_method was called. Aborting!\n");
     abort();
 }
```

rather than 

```cpp
const EC_METHOD *(*EC_GFp_nistp224_method)(void) = NULL;
```

**Comment by reaperhulk** [comment]
The diff between the 32-bit and 64-bit headers is as follows (green is 32-bit):

```diff
diff 1.0.2d_64/include/openssl/opensslconf.h 1.0.2d_32/include/openssl/opensslconf.h
13a14,16
> #ifndef OPENSSL_NO_EC_NISTP_64_GCC_128
> # define OPENSSL_NO_EC_NISTP_64_GCC_128
> #endif
64a68,70
> # if defined(OPENSSL_NO_EC_NISTP_64_GCC_128) && !defined(NO_EC_NISTP_64_GCC_128)
> #  define NO_EC_NISTP_64_GCC_128
> # endif
159c165
< #define DES_LONG unsigned int
---
> #define DES_LONG unsigned long
165c171
< #undef BN_LLONG
---
> #define BN_LLONG
170c176
< #define SIXTY_FOUR_BIT_LONG
---
> #undef SIXTY_FOUR_BIT_LONG
172c178
< #undef THIRTY_TWO_BIT
---
> #define THIRTY_TWO_BIT
184c190
< #undef BF_PTR
---
> #define BF_PTR
```

**Comment by haypo** [comment]
On Fedora 23, python -c 'import cryptography.hazmat.bindings._openssl' fails with:

```
ImportError: /usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so: undefined symbol: EC_GFp_nistp224_method
```

crytography was installed with 'pip install cryptography' (not from the source code).

Quick patch on the Git repository to workaround the issue:

```
diff --git a/src/_cffi_src/openssl/ec.py b/src/_cffi_src/openssl/ec.py
index e130a34..58da708 100644
--- a/src/_cffi_src/openssl/ec.py
+++ b/src/_cffi_src/openssl/ec.py
@@ -190,7 +190,6 @@ const EC_METHOD *EC_GFp_simple_method();
 const EC_METHOD *EC_GFp_mont_method();
 const EC_METHOD *EC_GFp_nist_method();
 
-const EC_METHOD *EC_GFp_nistp224_method();
 const EC_METHOD *EC_GFp_nistp256_method();
 const EC_METHOD *EC_GFp_nistp521_method();
 
@@ -372,7 +371,6 @@ static const long Cryptography_HAS_EC_1_0_1 = 1;
     defined(OPENSSL_NO_EC_NISTP_64_GCC_128)
 static const long Cryptography_HAS_EC_NISTP_64_GCC_128 = 0;
 
-const EC_METHOD *(*EC_GFp_nistp224_method)(void) = NULL;
 const EC_METHOD *(*EC_GFp_nistp256_method)(void) = NULL;
 const EC_METHOD *(*EC_GFp_nistp521_method)(void) = NULL;
 #else
diff --git a/src/cryptography/hazmat/bindings/openssl/_conditional.py b/src/cryptography/hazmat/bindings/openssl/_conditional.py
index 80a6dda..51c2e9e 100644
--- a/src/cryptography/hazmat/bindings/openssl/_conditional.py
+++ b/src/cryptography/hazmat/bindings/openssl/_conditional.py
@@ -154,7 +154,6 @@ CONDITIONAL_NAMES = {
     ],
 
     "Cryptography_HAS_EC_NISTP_64_GCC_128": [
-        "EC_GFp_nistp224_method",
         "EC_GFp_nistp256_method",
         "EC_GFp_nistp521_method",
     ],
```

**Comment by haypo** [comment]
More info on Fedora 23: libcrypto.so does not contains EC_GFp_nistp224_method, but has other EC_GFp_nistpXXX_method methods:

```
haypo@localhost$ objdump -T /usr/lib64/libcrypto.so |grep EC_GFp_nist
00000000000ef410 g    DF .text  0000000000000008  OPENSSL_1.0.1_EC EC_GFp_nistp256_method
00000000000f42a0 g    DF .text  0000000000000008  OPENSSL_1.0.1_EC EC_GFp_nistp521_method
00000000000f7580 g    DF .text  0000000000000008  OPENSSL_1.0.1_EC EC_GFp_nistz256_method
00000000000e1a70 g    DF .text  0000000000000008  OPENSSL_1.0.1_EC EC_GFp_nist_method
```

Source of the openssl package on Fedora 23: http://pkgs.fedoraproject.org/cgit/openssl.git/tree/?h=f23

http://pkgs.fedoraproject.org/cgit/openssl.git/tree/hobble-openssl?h=f23#n34 removes nistp224 per this commit:http://pkgs.fedoraproject.org/cgit/openssl.git/commit/hobble-openssl?h=f23&id=b3551463cafee86a63c60e294f754a8c5cddc37a

**Comment by reaperhulk** [comment]
This specific crash is now resolved in 1.1.2 although the underlying issue for the OS X problem still needs to be resolved upstream with homebrew.

(The undefined symbol issue on Fedora 23 is also resolved as of 1.1.2)
